File: src/com/sierrawireless/avphone/MainActivity.java
package com.sierrawireless.avphone;

import com.crashlytics.android.Crashlytics;
import io.fabric.sdk.android.Fabric;
import java.util.Date;

import net.airvantage.model.AvSystem;
import net.airvantage.utils.AvPhonePrefs;
import net.airvantage.utils.PreferenceUtils;
import android.app.ActionBar;
import android.app.ActionBar.Tab;
import android.app.ActionBar.TabListener;
import android.app.ActivityManager;
import android.app.ActivityManager.RunningServiceInfo;
import android.app.AlarmManager;
import android.app.FragmentTransaction;
import android.app.PendingIntent;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.content.SharedPreferences;
import android.content.SharedPreferences.OnSharedPreferenceChangeListener;
import android.os.Bundle;
import android.os.IBinder;
import android.preference.PreferenceManager;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentActivity;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentPagerAdapter;
import android.support.v4.view.ViewPager;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;

import com.sierrawireless.avphone.auth.Authentication;
import com.sierrawireless.avphone.auth.AuthenticationManager;
import com.sierrawireless.avphone.service.MonitoringService;
import com.sierrawireless.avphone.service.MonitoringService.ServiceBinder;
import com.sierrawireless.avphone.task.AsyncTaskFactory;
import com.sierrawireless.avphone.task.IAsyncTaskFactory;
import com.sierrawireless.avphone.task.SyncWithAvListener;
import com.sierrawireless.avphone.task.SyncWithAvResult;

/**
 * The main activity, in charge of displaying the tab view
 */
public class MainActivity extends FragmentActivity implements TabListener, LoginListener, AuthenticationManager,
        OnSharedPreferenceChangeListener, MonitorServiceManager, CustomLabelsManager, SyncWithAvListener {

    private static String LOGTAG = MainActivity.class.getName();

    private ViewPager viewPager;
    private TabsPagerAdapter tabsPageAdapter;
    private ActionBar actionBar;
    private AlarmManager alarmManager;
    private IAsyncTaskFactory taskFactory;
    private Authentication auth;
    private SharedPreferences prefs;

    boolean boundToMonitoringService = false;
    MonitoringService monitoringService;
    private MonitorServiceListener monitoringServiceListener = null;

    private CustomLabelsListener customLabelsListener = null;

    public void setCustomLabelsListener(CustomLabelsListener customLabelsListener) {
        this.customLabelsListener = customLabelsListener;
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Fabric.with(this, new Crashlytics());
        
        setContentView(R.layout.activity_main);

        prefs = PreferenceManager.getDefaultSharedPreferences(this);
        prefs.registerOnSharedPreferenceChangeListener(this);

        readAuthenticationFromPreferences();

        taskFactory = new AsyncTaskFactory(MainActivity.this);

        viewPager = (ViewPager) findViewById(R.id.pager);
        actionBar = getActionBar();
        tabsPageAdapter = new TabsPagerAdapter(getSupportFragmentManager());

        viewPager.setAdapter(tabsPageAdapter);
        actionBar.setHomeButtonEnabled(false);
        actionBar.setNavigationMode(ActionBar.NAVIGATION_MODE_TABS);

        alarmManager = (AlarmManager) this.getSystemService(Context.ALARM_SERVICE);

        // Adding Tabs
        Tab tab = actionBar.newTab().setText(getString(R.string.home_tab)).setTabListener(this);
        actionBar.addTab(tab);

        viewPager.setOnPageChangeListener(new ViewPager.OnPageChangeListener() {

            @Override
            public void onPageSelected(int position) {
                actionBar.setSelectedNavigationItem(position);
            }

            @Override
            public void onPageScrolled(int arg0, float arg1, int arg2) {
            }

            @Override
            public void onPageScrollStateChanged(int arg0) {
            }
        });

        if (isLogged()) {
            showLoggedTabs();

            if (isServiceRunning()) {
                connectToService();
            }
        } else {
            hideLoggedTabs();

            if (isServiceRunning()) {
                // The token is probably expired.
                // We stop the service since the "stop" button is not available anymore.
                this.stopMonitoringService();
            }
        }

    }

    private void readAuthenticationFromPreferences() {
        this.auth = PreferenceUtils.readAuthentication(this);
    }

    // Preferences

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {

        case R.id.action_settings:
            startActivity(new Intent(this, SettingsActivity.class));
            break;
        }

        return true;
    }

    @Override
    public void OnLoginChanged(boolean logged) {
    }

    private RunFragment runFragment;

    class TabsPagerAdapter extends FragmentPagerAdapter {

        public TabsPagerAdapter(FragmentManager fm) {
            super(fm);

        }

        @Override
        public Fragment getItem(int index) {

            switch (index) {
            case 0: {
                return makeHomeFragment();
            }
            case 1: {
                runFragment = (RunFragment) Fragment.instantiate(MainActivity.this, RunFragment.class.getName());
                return runFragment;
            }
            case 2:
                return makeConfigureFragment();
            }

            return null;
        }

        @Override
        public int getCount() {
            if (!isLogged()) {
                return 1;
            } else {
                return 3;
            }
        }

        protected HomeFragment makeHomeFragment() {
            HomeFragment fragment = (HomeFragment) Fragment
                    .instantiate(MainActivity.this, HomeFragment.class.getName());
            fragment.setTaskFactory(taskFactory);
            return fragment;
        }

        protected AvPhoneFragment makeConfigureFragment() {
            ConfigureFragment fragment = (ConfigureFragment) Fragment.instantiate(MainActivity.this,
                    ConfigureFragment.class.getName());
            fragment.setTaskFactory(taskFactory);
            return fragment;
        }

    }

    @Override
    public void onTabSelected(Tab tab, FragmentTransaction ft) {
        viewPager.setCurrentItem(tab.getPosition());
    }

    @Override
    public void onTabUnselected(Tab tab, FragmentTransaction ft) {
    }

    @Override
    public void onTabReselected(Tab tab, FragmentTransaction ft) {
    }

    public boolean isLogged() {
        return (this.auth != null && !this.auth.isExpired(new Date()));
    }

    @Override
    public void onAuthentication(Authentication auth) {

        this.auth = auth;

        saveAuthenticationInPreferences(auth);

        showLoggedTabs();

    }

    @Override
    public Authentication getAuthentication() {
        return this.auth;
    }

    private void saveAuthenticationInPreferences(Authentication auth) {
        PreferenceUtils.saveAuthentication(this, auth);
    }

    public void showLoggedTabs() {
        if (actionBar.getTabCount() == 1) {
            Tab runTab = actionBar.newTab().setText(getString(R.string.run_tab));
            actionBar.addTab(runTab.setTabListener(this));
            tabsPageAdapter.notifyDataSetChanged();
            actionBar.addTab(actionBar.newTab().setText(getString(R.string.configure_tab)).setTabListener(this));
            tabsPageAdapter.notifyDataSetChanged();
        }

        viewPager.setCurrentItem(1);
    }

    public void hideLoggedTabs() {
        if (actionBar.getTabCount() == 3) {
            actionBar.removeTabAt(2);
            tabsPageAdapter.notifyDataSetChanged();
            actionBar.removeTabAt(1);
            tabsPageAdapter.notifyDataSetChanged();
        }
        viewPager.setCurrentItem(0);
    }

    public void forgetAuthentication() {

        PreferenceUtils.resetAuthentication(this);

        this.auth = null;

        hideLoggedTabs();

        if (this.isServiceRunning()) {
            this.stopMonitoringService();
        }

    }

    @Override
    public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {

        if (PreferenceUtils.PREF_SERVER_KEY.equals(key) || PreferenceUtils.PREF_CLIENT_ID_KEY.equals(key)) {

            this.stopMonitoringService();
            this.forgetAuthentication();

        } else if (PreferenceUtils.PREF_PERIOD_KEY.equals(key) || PreferenceUtils.PREF_PASSWORD_KEY.equals(key)) {

            this.restartMonitoringService();

        } else if (key.indexOf("pref_data_custom") != -1) {

            if (this.customLabelsListener != null) {
                this.customLabelsListener.onCustomLabelsChanged();
            }
        }

    }

    public boolean isServiceRunning() {
        ActivityManager manager = (ActivityManager) this.getSystemService(Context.ACTIVITY_SERVICE);
        for (RunningServiceInfo service : manager.getRunningServices(Integer.MAX_VALUE)) {
            if (MonitoringService.class.getName().equals(service.service.getClassName())) {
                return true;
            }
        }
        return false;
    }

    private void connectToService() {
        Intent intent = new Intent(this, MonitoringService.class);
        boundToMonitoringService = this.bindService(intent, connection, Context.BIND_AUTO_CREATE);
    }

    private void disconnectFromService() {
        if (boundToMonitoringService) {
            this.unbindService(connection);
            boundToMonitoringService = false;
            if (monitoringServiceListener != null) {
                monitoringServiceListener.onServiceStopped(monitoringService);
            }
        }
    }

    ServiceConnection connection = new ServiceConnection() {

        @Override
        public void onServiceConnected(ComponentName arg0, IBinder binder) {
            Log.d(LOGTAG, "Connected to the monitoring service");
            monitoringService = ((ServiceBinder) binder).getService();

            if (monitoringServiceListener != null) {
                monitoringServiceListener.onServiceStarted(monitoringService);
            }

        }

        @Override
        public void onServiceDisconnected(ComponentName arg0) {
            Log.d(LOGTAG, "Disconnected from the monitoring service");
            boundToMonitoringService = false;
        }

    };

    public void sendAlarmEvent(boolean activated) {
        if (boundToMonitoringService && monitoringService != null) {
            monitoringService.sendAlarmEvent(activated);
        }
    }

    @Override
    public void startMonitoringService() {
        AvPhonePrefs avPrefs = PreferenceUtils.getAvPhonePrefs(this);

        Intent intent = new Intent(this, MonitoringService.class);
        intent.putExtra(MonitoringService.DEVICE_ID, DeviceInfo.getUniqueId(this));
        intent.putExtra(MonitoringService.SERVER_HOST, avPrefs.serverHost);
        intent.putExtra(MonitoringService.PASSWORD, avPrefs.password);

        PendingIntent pendingIntent = PendingIntent.getService(this, 0, intent, PendingIntent.FLAG_CANCEL_CURRENT);
        // registering our pending intent with alarm manager
        alarmManager.setRepeating(AlarmManager.ELAPSED_REALTIME_WAKEUP, 0, Integer.valueOf(avPrefs.period) * 60 * 1000,
                pendingIntent);

        connectToService();

    }

    @Override
    public void stopMonitoringService() {
        Intent intent = new Intent(this, MonitoringService.class);
        PendingIntent pendingIntent = PendingIntent.getService(this, 0, intent, PendingIntent.FLAG_CANCEL_CURRENT);
        alarmManager.cancel(pendingIntent);
        this.stopService(intent);

        disconnectFromService();

    }

    private void restartMonitoringService() {
        stopMonitoringService();
        startMonitoringService();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        disconnectFromService();
    }

    @Override
    public MonitoringService getMonitoringService() {
        return this.monitoringService;
    }

    public void setMonitoringServiceListener(MonitorServiceListener listener) {
        this.monitoringServiceListener = listener;
    }

    @Override
    public void onSynced(SyncWithAvResult result) {
        AvSystem system = result.getSystem();
        prefs.edit().putString("systemUid", system.uid).commit();
        prefs.edit().putString("systemName", system.name).commit();

        if (runFragment != null) {
            String systemUid = this.getSystemUid();
            String systemName = this.getSystemName();
            runFragment.setLinkToSystem(systemUid, systemName);
        } else {
            Log.w(LOGTAG, "RunFragment reference is null when onSynced is called");
        }

    }

    public String getSystemUid() {
        return prefs.getString("systemUid", null);
    }

    public String getSystemName() {
        return prefs.getString("systemName", null);
    }

}

File: mainActivity/src/main/java/com/sierrawireless/avphone/MainActivity.kt
package com.sierrawireless.avphone

import android.Manifest
import android.annotation.SuppressLint
import android.app.ActivityManager
import android.app.AlarmManager
import android.app.Fragment
import android.app.PendingIntent
import android.content.*
import android.content.SharedPreferences.OnSharedPreferenceChangeListener
import android.content.pm.PackageManager
import android.content.res.Configuration
import android.net.Uri
import android.os.Bundle
import android.os.IBinder
import android.preference.PreferenceManager
import android.support.v4.app.ActivityCompat
import android.support.v4.app.FragmentActivity
import android.support.v4.widget.DrawerLayout
import android.support.v7.app.ActionBarDrawerToggle
import android.util.Log
import android.view.MenuItem
import android.widget.ListView
import android.widget.Toast
import com.crashlytics.android.Crashlytics
import com.crashlytics.android.ndk.CrashlyticsNdk
import com.sierrawireless.avphone.auth.Authentication
import com.sierrawireless.avphone.auth.AuthenticationManager
import com.sierrawireless.avphone.service.MonitoringService
import com.sierrawireless.avphone.service.MonitoringService.ServiceBinder
import com.sierrawireless.avphone.task.*
import io.fabric.sdk.android.Fabric
import kotlinx.android.synthetic.main.activity_main.*
import net.airvantage.model.User
import net.airvantage.utils.PreferenceUtils
import java.util.*

/**
 * The main activity, handling drawer and Fragments
 */
class MainActivity : FragmentActivity(), LoginListener, AuthenticationManager, OnSharedPreferenceChangeListener, MonitorServiceManager, CustomLabelsManager, SyncWithAvListener {
    override var monitoringService: MonitoringService? = null
    private var objectName: String? = null

    private var alarmManager: AlarmManager? = null
    private var taskFactory: IAsyncTaskFactory? = null
    override var authentication: Authentication? = null
        set(value) {
            field = value
        }
    private var prefs: SharedPreferences? = null

    internal var boundToMonitoringService = false
    private var monitoringServiceListener: MonitorServiceListener? = null

    private var customLabelsListener: CustomLabelsListener? = null

    private var drawerToggle: ActionBarDrawerToggle? = null

    private var configureFragment: ConfigureFragment? = null
    private var homeFragment: HomeFragment? = null
    private var runFragment: ArrayList<RunFragment>? = null

    private var lastPosition = 0

    private var serviceSendData: Boolean? = false
    internal lateinit var objectsManager: ObjectsManager
    internal var user: User? = null

    override val isLogged: Boolean
        get() = this.authentication != null && !this.authentication!!.isExpired(Date())

    private var connection: ServiceConnection = object : ServiceConnection {

        override fun onServiceConnected(arg0: ComponentName, binder: IBinder) {
            Log.d(TAG, "Connected to the monitoring service")
            monitoringService = (binder as ServiceBinder).service

            if (monitoringServiceListener != null) {
                monitoringServiceListener!!.onServiceStarted(monitoringService!!)
            }

        }

        override fun onServiceDisconnected(arg0: ComponentName) {
            Log.d(TAG, "Disconnected from the monitoring service")
            boundToMonitoringService = false
        }

    }

    val systemUid: String?
        get() = prefs!!.getString("systemUid", null)

    var systemSerial: String?
        get() = prefs!!.getString(PREFERENCE_SYSTEM_SERIAL, null)
        @SuppressLint("DefaultLocale")
        set(serial) = prefs!!.edit().putString(PREFERENCE_SYSTEM_SERIAL, serial!!.toUpperCase()).apply()

    val systemName: String?
        get() = prefs!!.getString(PREFERENCE_SYSTEM_NAME, null)

    fun setUser(user: User) {
        this.user = user
    }

    private fun buildFragmentList(): ArrayList<MenuEntry> {
        val tmp = ArrayList<MenuEntry>()
        if (user != null) {
            tmp.add(MenuEntry("LOGGED AS", MenuEntryType.TITLE))
            tmp.add(MenuEntry(user!!.name!!, MenuEntryType.USER))
            tmp.add(MenuEntry(user!!.profile!!.name!!, MenuEntryType.USER))
            tmp.add(MenuEntry(user!!.company!!.name!!, MenuEntryType.USER))
            tmp.add(MenuEntry(user!!.server!!, MenuEntryType.USER))
        }
        tmp.add(MenuEntry("SIMULATED OBJECTS", MenuEntryType.TITLE))
        objectsManager.objects.mapTo(tmp) { MenuEntry(it.name!!, MenuEntryType.COMMAND) }
        tmp.add(MenuEntry(FRAGMENT_CONFIGURE, MenuEntryType.COMMAND))

        //tmp.add(FRAGMENT_SETTINGS);
        tmp.add(MenuEntry("NEED HELP", MenuEntryType.TITLE))
        tmp.add(MenuEntry(FRAGMENT_FAQ, MenuEntryType.COMMAND))
        tmp.add(MenuEntry("", MenuEntryType.TITLE))
        if (isLogged) {
            tmp.add(MenuEntry(FRAGMENT_LOGOUT, MenuEntryType.COMMAND))
        } else {
            tmp.add(MenuEntry(FRAGMENT_LOGIN, MenuEntryType.COMMAND))
        }
        return tmp
    }

    override fun setCustomLabelsListener(listener: CustomLabelsListener) {
        this.customLabelsListener = listener
    }

    //Get user
    private fun syncGetUser(auth: Authentication?) {


        val avPhonePrefs = PreferenceUtils.getAvPhonePrefs(this)


        val getUserTask = taskFactory!!.getUserTak(avPhonePrefs.serverHost!!, auth!!.accessToken!!)

        getUserTask.addProgressListener { result ->
            if (!result.isError) {
                user = result.user
                user!!.server = avPhonePrefs.serverHost
                loadMenu()

            }
        }

        val params = GetUserParams()

        getUserTask.execute(params)

    }


    override fun onCreate(savedInstanceState: Bundle?) {

        MainActivity.instance = this
        // Initialization og Object Manager
        objectsManager = ObjectsManager.getInstance()
        objectsManager.init(this)
        super.onCreate(savedInstanceState)
        Fabric.with(this, Crashlytics(), CrashlyticsNdk())



        setContentView(R.layout.activity_main)

        prefs = PreferenceManager.getDefaultSharedPreferences(this)
        prefs!!.registerOnSharedPreferenceChangeListener(this)

        //Auth is set here if exist
        readAuthenticationFromPreferences()

        taskFactory = AsyncTaskFactory(this@MainActivity)
        if (isLogged) {
            if (user == null) {
                syncGetUser(authentication)
            }
        }


        loadMenu()

        left_drawer.setOnItemClickListener { _, _, position, _ -> selectItem(position) }

        alarmManager = this.getSystemService(Context.ALARM_SERVICE) as AlarmManager

        drawerToggle = ActionBarDrawerToggle(this, drawer_layout, R.string.drawer_open,
                R.string.drawer_close)
        drawer_layout!!.addDrawerListener(drawerToggle!!)

        // Verify Permission

        when {
            ActivityCompat.checkSelfPermission(this, Manifest.permission.READ_PHONE_STATE) != PackageManager.PERMISSION_GRANTED -> ActivityCompat.requestPermissions(this@MainActivity, arrayOf(Manifest.permission.READ_PHONE_STATE), 1)
            ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED -> ActivityCompat.requestPermissions(this@MainActivity, arrayOf(Manifest.permission.ACCESS_COARSE_LOCATION), 2)
            ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED -> ActivityCompat.requestPermissions(this@MainActivity, arrayOf(Manifest.permission.ACCESS_FINE_LOCATION), 3)
        }
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        when (requestCode) {
            1 -> {
                if (grantResults.isNotEmpty()) {

                    if (grantResults[0] != PackageManager.PERMISSION_GRANTED) {

                        Log.d(TAG, "onRequestPermissionsResult: answer ko")

                        Toast.makeText(applicationContext, "Permission not granted please grant permission", Toast.LENGTH_LONG).show()
                    } else {
                        if (ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
                            ActivityCompat.requestPermissions(this@MainActivity, arrayOf(Manifest.permission.ACCESS_COARSE_LOCATION), 2)
                        } else if (ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED && ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
                            ActivityCompat.requestPermissions(this@MainActivity, arrayOf(Manifest.permission.ACCESS_FINE_LOCATION), 3)
                        }
                    }
                }
            }
            2 -> {
                if (grantResults.isNotEmpty()) {

                    if (grantResults[0] != PackageManager.PERMISSION_GRANTED) {

                        Log.d(TAG, "onRequestPermissionsResult: answer ko")

                        Toast.makeText(applicationContext, "Permission not granted please grant permission", Toast.LENGTH_LONG).show()
                    } else {
                        if (ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED && ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
                            ActivityCompat.requestPermissions(this@MainActivity, arrayOf(Manifest.permission.ACCESS_FINE_LOCATION), 3)
                        }
                    }
                }
            }
            3 -> {
                if (grantResults.isNotEmpty()) {
                    if (grantResults[0] != PackageManager.PERMISSION_GRANTED) {

                        Log.d(TAG, "onRequestPermissionsResult: answer ko")

                        Toast.makeText(applicationContext, "Permission not granted please grant permission", Toast.LENGTH_LONG).show()
                    }
                }
            }
            else -> super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        }
    }


    fun loadMenu() {
        FRAGMENT_LIST = buildFragmentList()


        val adapter = MenuAdapter(this, buildFragmentList())

        left_drawer.adapter = adapter

        left_drawer.choiceMode = ListView.CHOICE_MODE_SINGLE
        // Set the list's click listener
        left_drawer.invalidateViews()

        if (isLogged) {
            if (user == null) {
                syncGetUser(authentication)
            }

            unlockDrawer()
            if (isServiceRunning()) {
                connectToService()
            }

        } else {
            lockDrawer()
            if (isServiceRunning()) {
                // The token is probably expired.
                // We stop the service since the "stop" button is not available anymore.
                this.stopMonitoringService()
            }

        }
        goHomeFragment()


    }

    override fun onResume() {
        super.onResume()
        left_drawer.requestFocusFromTouch()
        left_drawer.setItemChecked(lastPosition, true)
        left_drawer.setSelection(lastPosition)
        left_drawer.refreshDrawableState()
        //drawerListView.setSelection();

    }

    override fun onPostCreate(savedInstanceState: Bundle?) {
        super.onPostCreate(savedInstanceState)
        drawerToggle!!.syncState()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return (drawerToggle!!.onOptionsItemSelected(item)
                // Or just go with defaults
                || super.onOptionsItemSelected(item))
    }

    override fun onConfigurationChanged(newConfig: Configuration) {
        super.onConfigurationChanged(newConfig)
        drawerToggle!!.onConfigurationChanged(newConfig)
    }

    private fun lockDrawer() {

        if (actionBar != null) {
            actionBar!!.setDisplayHomeAsUpEnabled(false)
            actionBar!!.setHomeButtonEnabled(false)
        }


        drawer_layout.setDrawerLockMode(DrawerLayout.LOCK_MODE_LOCKED_CLOSED)


        if (drawerToggle != null) {
            drawerToggle!!.isDrawerIndicatorEnabled = false
        }
    }

    private fun unlockDrawer() {

        if (actionBar != null) {
            actionBar!!.setDisplayHomeAsUpEnabled(true)
            actionBar!!.setHomeButtonEnabled(true)
        }


        drawer_layout.setDrawerLockMode(DrawerLayout.LOCK_MODE_UNLOCKED)

        if (drawerToggle != null) {
            drawerToggle!!.isDrawerIndicatorEnabled = true
        }
    }

    internal fun readAuthenticationFromPreferences() {
        this.authentication = PreferenceUtils.readAuthentication(this)
    }

    override fun onAuthentication(auth: Authentication) {

        this.authentication = auth
        unlockDrawer()

        saveAuthenticationInPreferences(auth)
    }

    private fun saveAuthenticationInPreferences(auth: Authentication) {
        PreferenceUtils.saveAuthentication(this, auth)
    }

    override fun forgetAuthentication() {

        lockDrawer()
        PreferenceUtils.resetAuthentication(this)
        this.authentication = null

        if (this.isServiceRunning()) {
            this.stopMonitoringService()
        }

    }

    override fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences, key: String) {

        if (PreferenceUtils.PREF_SERVER_KEY == key || PreferenceUtils.PREF_CLIENT_ID_KEY == key) {

            this.stopMonitoringService()
            this.forgetAuthentication()

        } else if (PreferenceUtils.PREF_PERIOD_KEY == key || PreferenceUtils.PREF_PASSWORD_KEY == key) {

            this.restartMonitoringService()

        } else if (key.contains("pref_data_custom")) {

            if (this.customLabelsListener != null) {
                this.customLabelsListener!!.onCustomLabelsChanged()
            }
        }

    }

     override fun isServiceRunning(): Boolean {
        //return serviceSendData;

        val manager = this.getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager

         @Suppress("DEPRECATION")
         return if (manager.getRunningServices(Integer.MAX_VALUE).any { MonitoringService::class.java.name == it.service.className }) serviceSendData!! else false

     }

    override fun isServiceStarted(name: String): Boolean {
        if (this.objectName == null) {
            return false
        }
        if (this.objectName != name) {
            return false
        }
        //return serviceSendData;

        val manager = this.getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager

        @Suppress("DEPRECATION")
        return manager.getRunningServices(Integer.MAX_VALUE).any { MonitoringService::class.java.name == it.service.className }

    }

    override fun oneServiceStarted(): Boolean {
        //return serviceSendData;

        val manager = this.getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager

        @Suppress("DEPRECATION")
        return manager.getRunningServices(Integer.MAX_VALUE).any { MonitoringService::class.java.name == it.service.className }

    }

    private fun connectToService() {
        val intent = Intent(this, MonitoringService::class.java)
        boundToMonitoringService = this.bindService(intent, connection, Context.BIND_AUTO_CREATE)
    }

    private fun disconnectFromService() {
        if (boundToMonitoringService) {
            this.unbindService(connection)
            boundToMonitoringService = false
            if (monitoringServiceListener != null) {
                monitoringServiceListener!!.onServiceStopped(monitoringService!!)
            }
        }
    }

    override fun sendAlarmEvent() {
        //if (boundToMonitoringService && monitoringService != null) {
        monitoringService!!.sendAlarmEvent()
        // }
    }

    override fun startMonitoringService(name: String) {
        this.objectName = name
        val avPrefs = PreferenceUtils.getAvPhonePrefs(this)

        val intent = Intent(this, MonitoringService::class.java)
        intent.putExtra(MonitoringService.DEVICE_ID, DeviceInfo.getUniqueId(this))
        intent.putExtra(MonitoringService.SERVER_HOST, avPrefs.serverHost)
        intent.putExtra(MonitoringService.PASSWORD, avPrefs.password)
        intent.putExtra(MonitoringService.CONNECT, false)
        intent.putExtra(MonitoringService.OBJECT_NAME, name)


        val pendingIntent = PendingIntent.getService(this, 0, intent,
                PendingIntent.FLAG_CANCEL_CURRENT)
        // registering our pending intent with alarm manager
        alarmManager!!.set(AlarmManager.ELAPSED_REALTIME_WAKEUP, 0, pendingIntent)

        connectToService()
    }

    override fun startSendData() {
        val avPrefs = PreferenceUtils.getAvPhonePrefs(this)

        val intent = Intent(this, MonitoringService::class.java)
        intent.putExtra(MonitoringService.DEVICE_ID, DeviceInfo.getUniqueId(this))
        intent.putExtra(MonitoringService.SERVER_HOST, avPrefs.serverHost)
        intent.putExtra(MonitoringService.PASSWORD, avPrefs.password)
        intent.putExtra(MonitoringService.CONNECT, true)

        val pendingIntent = PendingIntent.getService(this, 0, intent,
                PendingIntent.FLAG_CANCEL_CURRENT)
        // registering our pending intent with alarm manager
        alarmManager!!.setRepeating(AlarmManager.ELAPSED_REALTIME_WAKEUP, 5000,
                (Integer.valueOf(avPrefs.period)!! * 60 * 1000).toLong(), pendingIntent)

        // connectToService();
        serviceSendData = true

    }

    override fun stopSendData() {
        val intent = Intent(this, MonitoringService::class.java)
        val pendingIntent = PendingIntent.getService(this, 0, intent,
                PendingIntent.FLAG_CANCEL_CURRENT)
        alarmManager!!.cancel(pendingIntent)
        serviceSendData = false
    }

    override fun stopMonitoringService() {
        val intent = Intent(this, MonitoringService::class.java)
        val pendingIntent = PendingIntent.getService(this, 0, intent,
                PendingIntent.FLAG_CANCEL_CURRENT)
        alarmManager!!.cancel(pendingIntent)
        this.stopService(intent)

        disconnectFromService()
    }

    private fun restartMonitoringService() {
        stopMonitoringService()
        startMonitoringService(objectName!!)
    }

    public override fun onDestroy() {
        super.onDestroy()
        disconnectFromService()
    }


    override fun setMonitoringServiceListener(listener: MonitorServiceListener) {
        this.monitoringServiceListener = listener
    }

    @SuppressLint("DefaultLocale")
    override fun invoke(result: SyncWithAvResult) {

        val system = result.system ?: return
        prefs!!.edit().putString("systemUid", system.uid).apply()
        prefs!!.edit().putString(PREFERENCE_SYSTEM_NAME, system.name).apply()

        val user = result.user
        prefs!!.edit().putString(PREFERENCE_USER_UID, user!!.uid).apply()
        prefs!!.edit().putString(PREFERENCE_USER_NAME, user.name).apply()

        val deviceSerial = DeviceInfo.generateSerial(user.uid!!)
        prefs!!.edit().putString(PREFERENCE_SYSTEM_SERIAL, deviceSerial).apply()

        if (runFragment != null) {
            val systemUid = this.systemUid
            val systemName = this.systemName
            for (tmp in runFragment!!) {
                tmp.setLinkToSystem(systemUid, systemName)
            }
        } else {
            Log.w(TAG, "RunFragment reference is null when onSynced is called")
        }

    }

    /**
     * Swaps fragments in the main content view
     */
    private fun selectItem(position: Int) {

        val fragment = getFragment(position)
        val entry = FRAGMENT_LIST!![position]
        if (fragment == null) {
            //No item check if the position is valid
            Log.d(TAG, "selectItem: Fragment list " + entry.name)
            if (entry.name == FRAGMENT_FAQ) {
                val browserIntent = Intent(Intent.ACTION_VIEW, Uri.parse("http://doc.airvantage.net/alms/"))
                startActivity(browserIntent)
                left_drawer.setSelection(lastPosition)
                drawer_layout.closeDrawer(left_drawer)
            }
            return
        }
        if (entry.type == MenuEntryType.COMMAND) {

            // Insert the fragment by replacing any existing fragment
            fragmentManager
                    .beginTransaction()
                    .replace(R.id.content_frame, fragment)
                    .addToBackStack(null)
                    .commit()

            // Highlight the selected item, update the title, and close the drawer
            left_drawer.setItemChecked(position, true)
            title = entry.name
            left_drawer.setSelection(position)
            drawer_layout.closeDrawer(left_drawer)
            lastPosition = position
        }
    }

    fun goHomeFragment() {
        val position = FRAGMENT_LIST!!.size - 1
        val fragment = getFragment(position)
        // Insert the fragment by replacing any existing fragment
        fragmentManager
                .beginTransaction()
                .replace(R.id.content_frame, fragment)
                .addToBackStack(null)
                .commit()

        // Highlight the selected item, update the title, and close the drawer
        left_drawer.setItemChecked(position, true)
        title = FRAGMENT_LIST!![position].name
        left_drawer.setSelection(position)
        drawer_layout.closeDrawer(left_drawer)
        lastPosition = position

    }

    private fun initFragments(): Map<String, Fragment> {

        if (configureFragment == null) {
            configureFragment = ConfigureFragment()
            configureFragment!!.setTaskFactory(taskFactory!!)
            configureFragment!!.syncListener = this
        }

        if (homeFragment == null) {
            homeFragment = HomeFragment()
            homeFragment!!.setTaskFactory(taskFactory!!)
        }

        runFragment = ArrayList()

        var tmp: RunFragment
        for (obj in objectsManager.objects) {
            tmp = RunFragment()
            tmp.setTaskFactory(taskFactory!!)
            tmp.setObjectName(obj.name!!)
            runFragment!!.add(tmp)
        }


        val fragmentsMapping = HashMap<String, Fragment>()
        fragmentsMapping[FRAGMENT_CONFIGURE] = configureFragment!!
        if (isLogged) {
            fragmentsMapping[FRAGMENT_LOGOUT] = homeFragment!!
        } else {
            fragmentsMapping[FRAGMENT_LOGIN] = homeFragment!!
        }
        fragmentsMapping[FRAGMENT_SETTINGS] = SettingsActivity.SettingsFragment()
        for ((pos, obj) in objectsManager.objects.withIndex()) {
            fragmentsMapping[obj.name!!] = runFragment!![pos]
        }

        return fragmentsMapping
    }

    private fun getFragment(fragmentPosition: Int): Fragment? {
        val fragmentName = FRAGMENT_LIST!![fragmentPosition].name
        val fragmentMap = initFragments()
        return if (fragmentMap.containsKey(fragmentName)) fragmentMap[fragmentName] else null
    }

    companion object {

        private const val TAG = "MainActivity"
        private const val PREFERENCE_SYSTEM_NAME = "systemName"
        private const val PREFERENCE_SYSTEM_SERIAL = "systemSerial"
        private const val PREFERENCE_USER_NAME = "userName"
        private const val PREFERENCE_USER_UID = "userUid"

        private const val FRAGMENT_LOGOUT = "Logout"

        private const val FRAGMENT_LOGIN = "Login"
        private const val FRAGMENT_CONFIGURE = "Add/Modify/Delete"
        private const val FRAGMENT_SETTINGS = "Settings"
        private const val FRAGMENT_FAQ = "FAQ"

        @SuppressLint("StaticFieldLeak")
        internal lateinit var instance: MainActivity

        private var FRAGMENT_LIST: ArrayList<MenuEntry>? = null
    }

}
